### PR TITLE
Implement streaks

### PR DESCRIPTION
**Pull request by typermonkeyuser**
Hi,

I've added the concept of streaks.
I updated the profile page to show the user's current streak.
image

The streak will also affect the XP bonus, the larger the streak, the higher bonus.

I have updated the XP modifier to add the log10 calculation of the streak.
That means that a streak below 10 days will increase the modifier with 0.x, a streak below 100 will increase the modifier with 1.x, a streak below 1000 will increase the modifier with 2.x etc.

### CODE REVIEW

**Comment by Ferotiq** [comment]
I think the implementation could be better. Installing that package and calculating the streak on both the frontend and backend seems a bit wasteful.

Instead I propose I new user endpoint on the backend that just gets the streak of the user.

File: backend/package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "chalk": "4.1.2",
     "cors": "2.8.5",
     "cron": "1.8.2",
+    "date-streaks": "^1.2.1",
```
Comment (Ferotiq) [inline backend/package.json:24]: save exact

File: backend/src/api/controllers/result.ts [hunk 2]
```diff
@@ -34,6 +34,7 @@ import * as George from "../../tasks/george";
 import { getDailyLeaderboard } from "../../utils/daily-leaderboards";
 import AutoRoleList from "../../constants/auto-roles";
 import * as UserDAL from "../../dal/user";
+import { summary } from 'date-streaks';
```
Comment (Ferotiq) [inline backend/src/api/controllers/result.ts:37]: you need to run the pretty-fix script to fix all prettier related issues

File: backend/src/api/controllers/result.ts [hunk 3]
```diff
@@ -468,6 +469,20 @@ async function calculateXp(
       modifier += 0.1;
     }
   }
+  
+  // Increment modifier according to the user's current streak.
+  // Modifier is incremented with the log10 calculation of the current streak.
+  // That means that streaks greater than 10 days will increase modifier with 1.x,
+  // streaks greater than 100 will increase modifier with 2.x, greater than 1000 with 3.x, etc.
+  const allResults = await ResultDAL.getResults(uid);
```
Comment (Ferotiq) [inline backend/src/api/controllers/result.ts:477]: Getting all results might be a bit expensive and we could probably find a way to make this better.

Comment (Ferotiq) [reply backend/src/api/controllers/result.ts:477]: Also `getResults` returns only the last 1000 results and so the streak will be limited to some degree depending on how many tests a day the user does.

**Comment by fasttyperdog** [comment]
@Ferotiq I have moved the logic to the backend.
Maybe someone with more MongoDB knowledge than me could come up with a smart query to calculate the streak?

**Comment by fasttyperdog** [comment]
@Ferotiq I now store the streak on the user.
The expensive calculation is only necessary once, after that the streak is calculated based on the last result.

**Comment by fasttyperdog** [comment]
@Ferotiq Would you like to take an additional look, after my latest changes?

**Comment by Miodec** [comment]
I think it's fine to just avoid the init calculation and ignore any streaks before this feature is introduced

**Comment by fasttyperdog** [comment]
@Miodec Thanks, I have removed the code performing the initial calculation.

**Comment by fasttyperdog** [comment]
@Miodec Would you like to re-activate the tests?

**Comment by fasttyperdog** [comment]
@Miodec @Ferotiq Will one of you take an additional look when you have time? :)

**Comment by Miodec** [comment]
Why close?

**Comment by typermonkeyuser** [comment]
Mistake

**Comment by Miodec** [comment]
I don't like how you're getting the whole user doc everytime the profile is viewed. Plus it looks like if you view someone else's profile, their streak data will be put into the local db snapshot. 

You should get the streak data from the profile endpoint, and update local db snapshot with the data returned from adding result.

File: backend/__tests__/api/controllers/user.spec.ts [hunk 4]
```diff
@@ -52,5 +54,12 @@ describe("user controller test", () => {
         })
         .expect(409);
     });
+
```
Comment (Bruception) [inline backend/__tests__/api/controllers/user.spec.ts:57]: This is the wrong file for this test. There is a DAL test directory.

File: backend/src/api/controllers/user.ts [hunk 5]
```diff
@@ -487,3 +488,40 @@ export async function updateProfile(
 
   return new MonkeyResponse("Profile updated");
 }
+
+function isToday(someDate): boolean {
+  const today = new Date();
+  return (
+    someDate.getDate() == today.getDate() &&
```
Comment (Bruception) [inline backend/src/api/controllers/user.ts:495]: Use strict equality, also move these functions to utils/misc

**Comment by Miodec** [comment]
Waiting for you to address the issues raised by me and Bruce

**Comment by rizwanmustafa** [comment]
@typermonkeyuser any updates?

File: backend/__tests__/api/controllers/user.spec.ts [hunk 6]
```diff
@@ -52,5 +54,12 @@ describe("user controller test", () => {
...
+    it("updateStreak should update streak", async () => {
```
Comment (Bruception) [inline backend/__tests__/api/controllers/user.spec.ts:58]: This test should be moved to the dal spec

File: backend/src/api/controllers/user.ts [hunk 7]
```diff
@@ -4,13 +4,19 @@ import MonkeyError from "../../utils/error";
...
+  isYesterday,
+  sanitizeString,
+} from "../../utils/misc";
 import * as George from "../../tasks/george";
 import admin from "firebase-admin";
 import { deleteAllApeKeys } from "../../dal/ape-keys";
 import { deleteAllPresets } from "../../dal/preset";
 import { deleteAll as deleteAllResults } from "../../dal/result";
 import { deleteConfig } from "../../dal/config";
+import UserStreak = MonkeyTypes.UserStreak;
```
Comment (Bruception) [inline backend/src/api/controllers/user.ts:19]: I don't think we have to import this

File: backend/__tests__/api/controllers/user.spec.ts [hunk 8]
```diff
@@ -1,5 +1,7 @@
 import request from "supertest";
 import app from "../../../src/app";
+import * as UserDAL from "../../../src/dal/user";
```
Comment (Bruception) [inline backend/__tests__/api/controllers/user.spec.ts:3]: Unused importa

File: backend/__tests__/dal/user.spec.ts [hunk 9]
```diff
@@ -600,4 +605,25 @@ describe("UserDal", () => {
       },
     ]);
   });
+
+  it("updateStreak should update streak", async () => {
+    await UserDAL.addUser("testStack", "test email", "TestID");
+
+    const mockDate = new Date(1659749879000);
+    jest
```
Comment (Bruception) [inline backend/__tests__/dal/user.spec.ts:613]: Let’s restore this Mock

**Review by Bruception — APPROVED** [review]
Lgtm

**Pull request merged**